crates/credentials/src/vc.rs

```diff
@@ -127,25 +195,163 @@ impl VerifiableCredential {
+                if &named.id != iss {
+                    return Err(CredentialError::IssuerMismatch);
+                }
+            }
+        }
+
+        // optional expiration date check
+        if let (Some(exp), Some(vc_exp)) = (registered_claims.expiration, vc.expiration_date) {
+            if vc_exp == i64::default() {
+                vc.expiration_date = Option::from(exp);
+            } else if exp != vc_exp {
+                return Err(CredentialError::ExpirationDateMismatch);
+            }
+
+            let now = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs() as i64;
+
+            if now > vc.expiration_date.unwrap() {
+                return Err(CredentialError::CredentialExpired(
+                    "The verifiable credential has expired".to_string(),
+                ));
+            }
+        } else if registered_claims.expiration.is_none() && vc.expiration_date.is_some() {
+            return Err(CredentialError::MisconfiguredExpirationDate(
+                "VC has expiration date but no exp in registered claims".to_string(),
+            ));
+        }
+
+        // check for consistency and validity of expiration dates
+        if let (Some(exp), Some(vc_exp)) = (registered_claims.expiration, vc.expiration_date) {
+            // Update to registered expiration if current is default (0 or another default if defined)
+            if vc_exp == i64::default() {
+                vc.expiration_date = Some(exp);
+            } else if exp != vc_exp {
+                return Err(CredentialError::ExpirationDateMismatch);
+            }
+
+            // Check if the credential has expired
+            let current_time = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs() as i64;
+            if current_time > vc_exp {
+                return Err(CredentialError::CredentialExpired(
+                    "The verifiable credential has expired".to_string(),
+                ));
+            }
+        } else if registered_claims.expiration.is_none() && vc.expiration_date.is_some() {
+            return Err(CredentialError::MisconfiguredExpirationDate(
+                "VC has expiration date but no exp in registered claims".to_string(),
+            ));
+        }
+
+        validate_vc_data_model(&vc).map_err(CredentialError::ValidationError)?;
+
+        Ok(Arc::new(vc))
     }
 
     pub fn decode(jwt: &str) -> Result<Self, CredentialError> {
         let jwt_decoded = Jwt::decode::<VcJwtClaims>(jwt)?;
-
         Ok(jwt_decoded.claims.vc)
     }
 }
 
-// todo we should remove this altogether in the follow-up PR, but it would break bindings so leaving it for now
-pub async fn verify_vcjwt(jwt: &str) -> Result<Arc<VerifiableCredential>, CredentialError> {
-    let jwt_decoded = Jwt::verify::<VcJwtClaims>(jwt).await?;
-    Ok(Arc::new(jwt_decoded.claims.vc))
+fn validate_vc_data_model(
```

Comment:
I feel like this should be an instance method on the `impl VerifiableCredential` implementation, so maybe like 

```rust
impl VerifiableCredential {
    // ...
    pub fn validate(&self) -> Result<(), CredentialDataModelValidationError> {
        // ...
    }
}
```

Which then brings into question, perhaps we should add `*_vcjwt` to the `sign()`, `verify()` and `decode()` functions. So that would be `sign_vcjwt()`, `verify_vcjwt()` and `decode_vcjwt()`. Rust doesn't support function overloading FWIW.

Reply:
Well we've been down this path before

The thought was that the end user would get confused for
vc.validate()
vc.verify()

so the thought was to have a `verifyOptions` object, which you can have to only do validatePayload vs check specific verification vs all:
https://github.com/TBD54566975/web5-js/issues/425

Reply:
okay cool, I may make a follow up GitHub Issue for this matter -- specifically, for the endeavor of maximizing modularity here at the rust core layer I think it's fitting to offer every single one of these little verification/validation checks as its own function. my thought process is a barbell design, wherein `verify()` does everything, but then we expose public facing methods for each individual one, so if developers want to create their own set of verification rules then they string the atomic ones together